File: src/main/java/net/adventuregame/story/Quest.java
package net.adventuregame.story;

import java.util.function.Supplier;

public class Quest {
    private final String id;
    private final String description;
    private boolean completed = false;
    private Supplier<Boolean> completionCondition;

    public Quest(String id, String description, Supplier<Boolean> completionCondition) {
        this.id = id;
        this.description = description;
        this.completionCondition = completionCondition;
    }

    public String getId() {
        return id;
    }

    public String getDescription() {
        return description;
    }

    public boolean isCompleted() {
        if (completed) return true;
        if (completionCondition.get()) {
            completed = true;
            return true;
        }
        return false;
    }

    public void forceComplete() {
        completed = true;
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
    }

    public void setCompletionCondition(Supplier<Boolean> condition) {
        this.completionCondition = condition;
    }
}

File: src/main/kotlin/net/adventuregame/story/Quest.kt
package net.adventuregame.story

import java.util.function.Supplier

class Quest(val id: String?, val description: String?, private var completionCondition: Supplier<Boolean>) {
    private var completed = false

    var onComplete: (() -> Unit)? = null

    fun setOnCompleteRun(r: Runnable) {
        onComplete = { r.run() }
    }

    fun isCompleted(): Boolean {
        if (completed) return false  // already completed previously, so not “just completed”

        if (completionCondition.get()) {
            completed = true
            onComplete?.invoke() // only triggers once
            return true  // yes, just completed now
        }

        return false  // not completed yet
    }


    fun forceComplete() {
        completed = true
    }

    fun setCompleted(completed: Boolean) {
        this.completed = completed
    }

    fun setCompletionCondition(condition: Supplier<Boolean>) {
        this.completionCondition = condition
    }
}

